### DISCUSSION TITLE
Using autograd to optimize the resonant frequency of a cavity mode.

### DISCUSSION THREAD

**Discussion by tsgosain** [flexcompute/tidy3d#3359] [Q&A]
I am working on an inverse design problem where my loss function f(lambda/lambda target) depends on the cavity mode resonance wavelength and the target wavelength. Varying the size of my cavity changes the resonant wavelength of the mode. With the help of inverse designing, I want that size of the cavity for which the mode wavelength is equal to my target wavelength. Problem is to find the resonance wavelength I am using resonance finder function, but it is not compatible with autograd. Similarly, I tried taking the FFT of the time domain signal to get the E-field as a function of the wavelength, there autograd is not compatible with FFT. It seems like I have no option but to use numerical differentiation technique to calculate the derivatives with each parameter and it works but it is computationally very expensive. Is there any other way of solving this problem, where I can use autograd?

**Comment by momchil-flex** [comment]
Yeah we currently don't support adjoint for time-domain monitors, which is underlying all the limitations you're hitting. Unfortunately I can't think of a good way to optimize the cavity resonance frequency directly, but maybe you can achieve some results if you manage to phrase your objective function as something that can be recorded by frequency-domain monitors.

For example, say you're optimizing a cavity for Purcell emission. We know that for a single-mode cavity the Purcell factor is something like Q/V. However, adjoint method for the Q is not supported. But if you instead just maximize the flux emitted by the dipole source, and your starting point looks like a cavity, then the optimization is likely to find some optimal Q/V ratio without explicitly computing the Q or the V.

This is just for illustrative purposes, I don't know whether your problem yields to some such treatment.

**Reply by tsgosain** [reply]
What I was thinking was that at resonance the intensity will be maximum. I have a field monitor at one frequency which is my target frequency. Now when I change the size of my cavity the resonant frequency is going to shift hence the intensity at my target frequency will decrease as the intensity will be max at some other freq. So, if I define my objective function to be just intensity then can we say that at that cavity size where the mode frequency is exactly equal to the target frequency then the gradient of the objective function will be zero at that cavity size?

**Reply by momchil-flex** [reply]
This is generally correct but the problem is that if you have a high Q, and your monitor frequency is many linewidths away from the resonance frequency, it's in the exponentially suppressed "tails" of the mode and the gradient will be extremely small. So this will work well only if your starting frequency is reasonably within the cavity linewidth.

What's the ball park Q you're wokring with? In general, inverse design in high-Q structures is one of the most challenging applications to get right. You could for example try to ameliorate this by looking at multiple monitor frequencies, hoping to catch the resonance and define your OF based on that, but if it's too high-Q it will require too many frequency points.

**Reply by tsgosain** [reply]
I am designing a nanobeam cavity and the Q-factor is approximately 500, so the high Q problem will not be there. I have a field monitor placed along x-y plane at my resonant frequency. So, I am summing up the intensity at each spatial point to get the total intensity at the desired frequency. Now I have kept the cavity size to be such that my mode frequency is equal to the target frequency, and I can see that the total intensity has maximized as compared to other cavity size, but the value of the gradient is not coming to be zero rather I got highly negative value of the gradient. I don't understand this. Should the gradient be zero?

**Reply by momchil-flex** [reply]
Yeah 500 might work.

I am not sure about what you're observing. Maybe try to make the monitor smaller, like just around the center of the cavity? When your adjoint simulation depends on a FieldMonitor, it places a custom current source that spans the entire field monitor span. Maybe it's not great to inject currents all over your resonator.

**Reply by tsgosain** [reply]
To excite the cavity, I am using a dipole source with a particular polarization and to avoid dipole transients I use apodization so that the monitors capture only the cavity response. In adjoint simulation there is an adjoint field monitor which measures the E field components at different frequency but no apodization was applied there. Can that be a problem due to which I am getting such weird values of gradient as the gradient depends on the overlap of the forward and adjoint field.

**Reply by tsgosain** [reply]
Our objective function should always depend on the electric field for example like intensity, flux etc ? what if in my case if I consider the objective function to be (1 - lambda/lambda_target)**2 where the function minimizes when the cavity mode is equal to the target wavelength?

**Reply by momchil-flex** [reply]
> To excite the cavity, I am using a dipole source with a particular polarization and to avoid dipole transients I use apodization so that the monitors capture only the cavity response. In adjoint simulation there is an adjoint field monitor which measures the E field components at different frequency but no apodization was applied there. Can that be a problem due to which I am getting such weird values of gradient as the gradient depends on the overlap of the forward and adjoint field.

Yeah that's a good question, I was wondering about apodization too. Note that I'm just saying things out of my own intuition here, I have not worked on this problem and I don't know anyone on the team who has. That said, my first comment would be that the optimization should work *without* apodization too, in other words, if you just look at the field intensity of the monitor without apodization, I would think you should still see it maximized when the monitor frequency matches the cavity frequency. It's good to use apodization when you want to capture the cavity mode independent from the source, but if you're just looking for the resonance, the injection will be maximized at resonance anyway.

If you place a small-ish adjoint monitor around the original dipole, then the adjoint simulation will also be something like a localized injection, which I think should be ok. The only thing is that the normalization of the gradient in what you're doing, and in general, will be kind of arbitrary. This is actually making me wonder about your statement

> but the value of the gradient is not coming to be zero rather I got highly negative value of the gradient

I wonder if it's just bad normalization - can you check the gradient at several different frequencies? Is the gradient still minimized around the cavity freqeuncy and you just need to rescale it by some large factor during the optimization, but apart from that it works?

> Our objective function should always depend on the electric field for example like intensity, flux etc ? what if in my case if I consider the objective function to be (1 - lambda/lambda_target)**2 where the function minimizes when the cavity mode is equal to the target wavelength?

I guess this doesn't sound very well-defined, what is `lambda`? I think this suggestion is coming back to being able to differentiate through the time-domain monitor and the ResonanceFinder, which will spit out the `lambda`, but this is not supported.

**Reply by tsgosain** [reply]
Please correct me if I am wrong. I should place a tiny field monitor at the dipole's position and hence as a result the adjoint source would be tiny current source in the adjoint simulation. Regarding the normalization of the gradient, I have not done that. I just used the auto grad function to calculate the gradient and value of my objective function. Should the direction of the current source be same as the polarization of the dipole? Also, I am not able to see the results of the adjoint simulation. Somehow the "Preparing files for visualizing simulation results" always get stuck at 98 percent for all my adjoint simulation files. Why is that happening?

**Reply by momchil-flex** [reply]
> Should the direction of the current source be same as the polarization of the dipole?

This will be automatically handled by agutograd

> Also, I am not able to see the results of the adjoint simulation. Somehow the "Preparing files for visualizing simulation results" always get stuck at 98 percent for all my adjoint simulation files. Why is that happening?

Hm not sure exactly but maybe adjoint sims are not meant to be displayed in GUI. But if you have the task id, you should be able to examine it in python with `sim_data_adj = web.load(task_id)`.

> Regarding the normalization of the gradient, I have not done that. I just used the auto grad function to calculate the gradient and value of my objective function.

Yeah I think it's worth a check just in case there's just some large normalization factor that is arbitrary, but otherwise things are correct (like an optimization will work well)

**Reply by tsgosain** [reply]
I tried downloading the adjoint simulation files from the server and it gave me an error message "Failed to download the simulation data file from the server. Please confirm that the task was successfully run." even though the simulation was run successfully. This is happening with all the adjoint simulation files. The forward simulations files are downloadable.

**Reply by tsgosain** [reply]
I have tried one approach to achieve the target resonant wavelength. I have placed a field monitor at one end of my cavity, and I am measuring the flux, and let the target wavelength is 780 nm and at the optimized values of cavity parameters the flux will get maximized at 780 nm whereas when the parameters values shift from the optimized values and the resonant frequency of the mode is shifted away from the target wavelength and hence the flux at the target wavelength will get reduced. The field monitor measures the flux at the target wavelength for different cavity parameters and that cavity parameter for which the flux is maximized is the optimized parameter. I have normalized the flux with the dipole power at the target wavelength. But what I have seen is that the gradient becomes maximum at the optimized parameter value. So while updating the parameter instead of multiplying the learning rate withe gradient, I am multiplying it with 1/gradient and since I am going towards point of maxima increase of descent, I am using ascent (x_n+1 = x_n + LR*1/gradient).  
<img width="571" height="436" alt="Flux_550" src="https://github.com/user-attachments/assets/3015ce75-33af-4277-a909-347994c73120" />
<img width="587" height="436" alt="Grad_550" src="https://github.com/user-attachments/assets/d3693d2c-7480-45d5-a9e2-1abe428dd901" />
<img width="582" height="436" alt="Inv_Grad_550" src="https://github.com/user-attachments/assets/1f4d0888-f483-4da0-8d97-60bd332f4e68" />
<img width="571" height="436" alt="Obj_550" src="https://github.com/user-attachments/assets/68bdcf1e-a7a1-4a47-bb1b-69075e0b1880" />
This is true for other values of the cavity parameters.

**Comment by momchil-flex** [comment]
Ah, that's interesting. I think the crux of the matter is visible here. *Exactly* on resonance, there is some point where the gradient is zero. However, because of the sharp peak, the gradient becomes very large when you're slightly off-resonance.

<img width="587" height="436" alt="image" src="https://github.com/user-attachments/assets/f37eec55-4f22-4566-a152-fca420178f8e" />

Your workaround makes sense. This might also be a know issue in engineering/optimization problems and there might be standard ways to handle it? Maybe a chat with your favorite AI tool can improve on what you've got, unless you think it's good enough already.
